tests/Obsidian/Cache.test.ts

```diff
@@ -285,4 +292,178 @@ describe('cache', () => {
+
+        expect(task.parent).toEqual(null);
+
+        expect(task.children.length).toEqual(1);
+
+        const listItem = task.children[0];
+
+        expect(listItem.parent).toEqual(task);
+        expect(listItem.children.length).toEqual(0);
+        expect(listItem.originalMarkdown).toEqual('    - child list item');
+    });
+
+    it('should read parent task, child listItem and grandchild task', () => {
+        const tasks = readTasksFromSimulatedFile(inheritance_task_listitem_task);
+        expect(inheritance_task_listitem_task.fileContents).toMatchInlineSnapshot(`
+            "- [ ] parent task
+                - child list item
+                    - [ ] grandchild task
+            "
+        `);
+
+        expect(tasks.length).toEqual(2);
+
+        const [parentTask, grandChildTask] = tasks;
+
+        expect(parentTask.parent).toEqual(null);
+
+        expect(parentTask.children.length).toEqual(1);
+
+        const childListItem = parentTask.children[0];
+
+        expect(childListItem.parent).toEqual(parentTask);
+        expect(childListItem.children).toEqual([grandChildTask]);
+        expect(childListItem.originalMarkdown).toEqual('    - child list item');
+
+        expect(grandChildTask.parent).toEqual(childListItem);
+        expect(grandChildTask.children).toEqual([]);
+    });
+
+    it('should read parent task, two child listItems and 3 grandchild tasks', () => {
+        const tasks = readTasksFromSimulatedFile(inheritance_task_2listitem_3task);
+        expect(inheritance_task_2listitem_3task.fileContents).toMatchInlineSnapshot(`
+            "- [ ] parent task
+                - child list item 1
+                    - [ ] grandchild task 1
+                    - [ ] grandchild task 2
+                - child list item 2
+                    - [ ] grandchild task 3
+            "
+        `);
+
+        expect(tasks.length).toEqual(4);
+
+        const [parentTask, grandChildTask1, grandChildTask2, grandChildTask3] = tasks;
+
+        expect(parentTask.parent).toEqual(null);
+
+        expect(parentTask.children.length).toEqual(2);
+
+        const [childListItem1, childListItem2] = parentTask.children;
+
+        expect(childListItem1.parent).toEqual(parentTask);
+        expect(childListItem1.children).toEqual([grandChildTask1, grandChildTask2]);
+        expect(childListItem1.originalMarkdown).toEqual('    - child list item 1');
+
+        expect(grandChildTask1.parent).toEqual(childListItem1);
+        expect(grandChildTask1.children).toEqual([]);
+
+        expect(grandChildTask2.parent).toEqual(childListItem1);
+        expect(grandChildTask2.children).toEqual([]);
+
+        expect(childListItem2.parent).toEqual(parentTask);
+        expect(childListItem2.children).toEqual([grandChildTask3]);
+        expect(childListItem2.originalMarkdown).toEqual('    - child list item 2');
+
+        expect(grandChildTask3.parent).toEqual(childListItem2);
+        expect(grandChildTask3.children).toEqual([]);
```

Comment:
I don't currently have enough time to review all these new tests sufficiently thoroughly, unfortunately

@ilandikov, would you consider seeing if you can make the behaviour of these tests visible via an approval test that shows the markdown input, and the parent/child `ListItem` structure?

You might need to make `getTasksFromFileContent2() return` all the root items, including `ListItem`s, in order for that test to work.

Reply:
> if you can make the behaviour of these tests visible

Or rather, make the behaviour of `getTasksFromFileContent2()` visible... 😄

Reply:
> @ilandikov, would you consider seeing if you can make the behaviour of these tests visible via an approval test that shows the markdown input, and the parent/child `ListItem` structure?

Do you need this to merge this PR? If yes, I will be making a PR to @stasmarkin first so that the changes will be reflected here. Otherwise I can do a separate PR after this one is merged.

Reply:
> > @ilandikov, would you consider seeing if you can make the behaviour of these tests visible via an approval test that shows the markdown input, and the parent/child `ListItem` structure?
> 
> Do you need this to merge this PR? If yes, I will be making a PR to @stasmarkin first so that the changes will be reflected here. Otherwise I can do a separate PR after this one is merged.

I do prefer to properly understand tests before merging a PR, so yes please.

Reply:
@claremacrae does this look better? I prefer to verify separately before making PR to @stasmarkin =)

https://github.com/stasmarkin/obsidian-tasks/compare/refactor-store-listitems...ilandikov:obsidian-tasks:test-improve-hierarchy-testing?expand=1

Reply:
Ah no, sorry, I really did mean an [Approval Test](https://publish.obsidian.md/tasks-contributing/Testing/Approval+Tests).

Reply:
I will "pack" this into an approval test for sure but first I wanted to make sure that the printers were ok in an inline snapshot =)

Reply:
Ah right - great.

One comment is that now, is the code actually testing whether the correct object type is being created.
By writing out the original markdown line, would the code still pass if they were all created as `ListItem` instances?

Reply:
@claremacrae added the types, please take a look, if ok for you, I will migrate to something like `verifyHierarchies(simulatedFile: SimulatedFile)` that will be printing:

- markdown of the file
- hierarchy with `printHierarchy()`

I guess there is no point in printing total number of `Task` objects now since the types are included in the snapshot, so the lines like `expect(tasks.length).toEqual(2);` will just vanish.

I also think the whole helper functions' suite can be exported to something like `RelationshipHelpers.ts` (Nice name, hah).

Reply:
On the other hand, I would appreciate if we could work on the migration to approvals together if that is ok for you. I have a draft but not sure whether it is good enough. So I suggest accepting this with inline snapshots for now. Let me know what you think =)

Reply:
Sorry, I'm a bit lot in that conversation. Is there anything I can do about that in this PR?

Reply:
> Sorry, I'm a bit lot in that conversation. Is there anything I can do about that in this PR?

No, nothing, thanks.

It started as a hope that we could avoid adding any more synonyms than:

> - inheritance
> - parent/child/grandchild
> - and now relationships...

But I've just been looking at the Obsidian docs, and they talk about [Nesting lists](https://help.obsidian.md/Editing+and+formatting/Basic+formatting+syntax#Nesting+lists) and `indenting` - so they seem like the way to go, in code and (eventually) documentation.